fix: make compatibility versions test use dynamic n-1 branch for test (from hardcoded release-1.31)

aaron-prindle opened the pull request:
Fixes #33573 

This PR modifies `experiment/compatibility-versions/e2e-k8s-compatibility-versions.sh` to fetch the latest release-X.Y branch from https://github.com/kubernetes/kubernetes.git and use that as the n-1 version for compatibility versions testing.  This way this test is dynamic and updates the n-1 version as new releases are made - updating automatically over time vs the current hardcoded `release-1.31` value currently present in the test.

aaron-prindle commented:
/cc @BenTheElder

BenTheElder commented on experiment/compatibility-versions/e2e-k8s-compatibility-versions.sh:
```diff
@@ -327,10 +328,16 @@ main() {
-  # Clone the specific Kubernetes release branch
-  clone_kubernetes_release
+  # Clone the previous versions Kubernetes release branch
+  # TODO(aaron-prindle) extend the branches to test from n-1 -> n-1..3 as more k8s releases are done that support compatibility versions
```
We should also think about which branches we're trying to test, if what we're trying to test is the api-server compatibility in `master`, it _might_ better suit us to use the latest stable tagged patch release in the previous branch for the other components. In practice there's not usually a big distinction anyhow though.

BenTheElder replied:
Also, if we did that we could fetch binaries instead of compiling both branches from source.

aaron-prindle replied:
Makes sense, given that it doesn't seem to big a distinction atm perhaps we can keep this as is for now?  I created https://github.com/kubernetes/test-infra/issues/33594 to track using precompiled binaries

BenTheElder replied:
Yeah, I just wanted to point that out, with the precompiled builds we have a bunch of different marker files for the subtle variations depending on what you're actually focusing on testing and how stable you want the other end of the skew.

For now 🤷, let's just get it working, but later we'll want to decide.

BenTheElder commented on experiment/compatibility-versions/e2e-k8s-compatibility-versions.sh:
```diff
@@ -286,17 +283,19 @@ run_tests() {
-  git clone --single-branch --branch "${KUBE_RELEASE_BRANCH}" https://github.com/kubernetes/kubernetes.git
+get_latest_release_version() {
+    # Fetch all branch names
+    git ls-remote --heads https://github.com/kubernetes/kubernetes.git | \
```
we should probably prefer using local git state?

we can list branches or tags locally and determine this, and it will work in offline development

BenTheElder replied:
actually, further, to make this work with N branches, it should look at something like `git describe`, get the current release branch from that and then compute the N-... branch.

Because we'll want to be running this on each release going forward and those can't just assume the target is the latest release at all.

aaron-prindle replied:
Does the k8s repo that Prow pulls have all of the other branches available?  If so, definitely +1 - just want to confirm that.  We pass in `    base_ref: master` as part of the repo config so wasn't sure if all branches were pulled down.

BenTheElder commented on experiment/compatibility-versions/e2e-k8s-compatibility-versions.sh:
```diff
@@ -327,10 +328,16 @@ main() {
+  # Clone the previous versions Kubernetes release branch
+  # TODO(aaron-prindle) extend the branches to test from n-1 -> n-1..3 as more k8s releases are done that support compatibility versions
+  export PREV_RELEASE_BRANCH="release-${EMULATED_VERSION}"
+  git clone --single-branch --branch "${PREV_RELEASE_BRANCH}" https://github.com/kubernetes/kubernetes.git "${PREV_RELEASE_BRANCH}"
```
Do we really need another full clone? (also if we're going to clone, we should add some performance related flags, cloning kubernetes in full is really expensive with the 10 years of history)

If we need to use git, let's use a worktree. But I think we can fetch binaries for the old release and for the new release we should already have the checkout in order, presumably.

aaron-prindle replied:
+1, I'm familiar with how to pull branch with minimal history (`--depth=1`) which I've changed this to use now.  Can you explain how a worktree would help in this context or give an example for what you are suggesting with `If we need to use git, let's use a worktree.`?  Thanks

BenTheElder replied:
> +1, I'm familiar with some of the flags to pull branch with minimal history (--depth=1, --single-branch) which I've changed this to use now. Did you have any other suggestions for flags there?

`--depth=1` will break `git describe`, in prow we use `--filter=blob:none` instead

aaron-prindle replied:
Ah got it, I've updated this to use `--filter=blob:none` now:
```
  git clone --filter=blob:none --single-branch --branch "${PREV_RELEASE_BRANCH}" https://github.com/kubernetes/kubernetes.git "${PREV_RELEASE_BRANCH}"

```

BenTheElder commented:
TLDR I think:
- let's determine which release we're on currently by using hack/print-workspace-status.sh (so single source of truth about kube versioning from the main repo), that way this script "just works" for 1.32, 1.33, 1.34, ...
- then we can compute the N-1 or N-2 release minor version (maybe we should make the constant minor back-compat offset a variable with a default of 1 or 2?)
- we can then do `kind build node image v1.N-1.(latest)` (TODO: we can get latest from `git tag | grep 'v1.N-1.'` ?)
- create kind cluster from this image, ensure --compat-version is set to N-1
- then we can build and patch in api-server from the current branch
- run tests from N-1 ? (we can also download these the same as kind downloading the component release binaries)

aaron-prindle commented:
> - let's determine which release we're on currently by using hack/print-workspace-status.sh (so single source of truth about kube versioning from the main repo), that way this script "just works" for 1.32, 1.33, 1.34, ... (instead of floating with the current latest)
Ah I didn't know about hack/print-workspace-status.sh, thanks!   I've updated the logic here to use that script vs looking at the git remote information.  

> - then we can compute the N-1 or N-2 release minor version (maybe we should make the constant minor back-compat offset a variable with a default of 1 or 2?)
Currently I've updated this PR to compute only the N-1 release minor version as v1.31-v1.32 are the only versions support for Compatibility Versions currently

Currently the PR is focused on fixing #33573.  I've created #33849 to track the issue with the kubelet version being incorrect.  I will address the below points in the PR to fix #33849 if that is ok.  Can also do them both in the PR, currently my plan is to do them in 2 PRs though.
> - we can then do kind build node image v1.N-1.(latest) (TODO: we can get latest from git tag | grep 'v1.N-1.' ?)
> - create kind cluster from this image, ensure --compat-version is set to N-1
> - then we can build and patch in api-server from the current branch
> - run tests from N-1 ? (we can also download these the same as kind downloading the component release binaries)

BenTheElder reviewed (commented):
/lgtm
/approve
